quic/s2n-quic-core/src/buffer/reassembler.rs

```diff
@@ -158,115 +166,230 @@ impl Reassembler {
     /// Pushes a slice at a certain offset
     #[inline]
     pub fn write_at(&mut self, offset: VarInt, data: &[u8]) -> Result<(), Error> {
-        // create a request
-        let request = Request::new(offset, data)?;
-        self.write_request(request)?;
+        let mut request = Request::new(offset, data, false)?;
+        self.write_reader(&mut request)?;
         Ok(())
     }
 
     /// Pushes a slice at a certain offset, which is the end of the buffer
     #[inline]
     pub fn write_at_fin(&mut self, offset: VarInt, data: &[u8]) -> Result<(), Error> {
-        // create a request
-        let request = Request::new(offset, data)?;
-
-        // compute the final offset for the fin request
-        let final_offset = request.end_exclusive();
+        let mut request = Request::new(offset, data, true)?;
+        self.write_reader(&mut request)?;
+        Ok(())
+    }
 
-        // make sure if we previously saw a final size that they still match
-        //= https://www.rfc-editor.org/rfc/rfc9000#section-4.5
-        //# Once a final size for a stream is known, it cannot change.  If a
-        //# RESET_STREAM or STREAM frame is received indicating a change in the
-        //# final size for the stream, an endpoint SHOULD respond with an error
-        //# of type FINAL_SIZE_ERROR; see Section 11 for details on error
-        //# handling.
-        if let Some(final_size) = self.final_size() {
-            ensure!(final_size == final_offset, Err(Error::InvalidFin));
+    #[inline]
+    pub fn write_reader<R>(&mut self, reader: &mut R) -> Result<(), Error<R::Error>>
+    where
+        R: Reader + ?Sized,
+    {
+        // Trims off any data that has already been received
+        reader.skip_until(self.current_offset())?;
+
+        // store a snapshot of the cursors in case there's an error
+        let snapshot = self.cursors;
+
+        self.check_reader_fin(reader)?;
+
+        if let Err(err) = self.write_reader_impl(reader) {
+            use core::any::TypeId;
+            if TypeId::of::<R::Error>() != TypeId::of::<core::convert::Infallible>() {
```

Comment:
how would an error be returned if it was `Infallible`?

Reply:
It wouldn't. Which is exactly why I check to see if the error is not infallible to know if I need to roll the changes back. I wanted to remove this check but it ended up being needed to avoid the branch, unfortunately. The perf ended up being about 5% worse without it.